backpack/utils/conv_transpose.py

```diff
@@ -23,6 +24,53 @@ def get_weight_gradient_factors(input, grad_out, module, N):
     return X, dE_dY
 
 
+def extract_weight_diagonal(module, input, grad_output, N):
+    """
+    input must be the unfolded input to the convolution (see unfold_func)
+    and grad_output the backpropagated gradient
+    """
+
+    out_channels = module.weight.shape[0]
+    in_channels = module.weight.shape[1]
+    k = module.weight.shape[2:]
+
+    M = input.shape[0]
+    output_shape = module(module.input0).shape
+    spatial_out_size = output_shape[2:]
+    spatial_out_numel = spatial_out_size.numel()
+
+    input_reshaped = input.reshape(M, -1, spatial_out_numel)
+
+    if N == 1:
+        grad_output_viewed = grad_output
+    elif N == 2:
+        grad_output_viewed = eingroup("v,n,c,h,w->v,n,c,hw", grad_output)
+    elif N == 3:
+        grad_output_viewed = eingroup("v,n,c,d,h,w->v,n,c,dhw", grad_output)
+    else:
+        raise ValueError("{}-dimensional Conv. is not implemented.".format(N))
+
+    AX = einsum("nkl,vnml->vnkm", (input_reshaped, grad_output_viewed))
+    weight_diagonal = (AX ** 2).sum([0, 1]).transpose(0, 1)
+    weight_diagonal = weight_diagonal.reshape(in_channels, out_channels, *k)
+    return weight_diagonal.transpose(0, 1)
+
+
+def extract_bias_diagonal(module, sqrt, N):
+    """
+    `sqrt` must be the backpropagated quantity for DiagH or DiagGGN(MC)
+    """
+    V_axis, N_axis = 0, 1
+
+    if N == 1:
+        bias_diagonal = (einsum("vncl->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+    elif N == 2:
+        bias_diagonal = (einsum("vnchw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+    elif N == 3:
+        bias_diagonal = (einsum("vncdhw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
```

Comment:
Can you pull out the `einsum, **2, .sum([...])` to reduce duplication? Also, raise a `ValueError` if `N` is not `1,2,3`